Pull request for drawing signature field

nathando opened the pull request:
- Add controller for signature field, including buttons for save and reset
- Add library and style

nathando committed:
- Add controller for signature field, including buttons for save and reset
- Add library and style

yashodhank commented:
@nathando Thanks for pull request. Will be able to post screenshots as well here? it will be helpful for reviewer.

nathando commented:
Sure, here is the screenshot of the current state. As mentioned [here](https://discuss.erpnext.com/t/drawn-signature-html-field/6824/23), this commit wasn't able to save the image on server yet but the base64 string was already captured. 

![signature field](https://cloud.githubusercontent.com/assets/1678810/21024653/4da16bb2-bdc0-11e6-9734-2fbe6a672d7a.png)

nathando commented:
Hi, so do you have any advise to complete this feature. I can have the base64 str on client but lost it when saving. Wanna complete this as fast as possible

nathando committed:
Add DB mapping for Signature field -> fix bug not saving
Add empty image, and auto load image when loading form
Add option for Signature field type on Doctype form

nathando commented:
@yashodhank Hi there, i've fixed the problem with the previous version. Could you help to review now ? Thanks!

nathando committed:
Merge branch 'develop' into develop

ayamchiken committed:
Merge branch 'develop' of https://github.com/nathando/frappe into develop

Conflicts:
	frappe/public/build.json

netchampfaris commented on frappe/public/css/form.css:
```diff
@@ -444,9 +444,22 @@ select.form-control {
 .form-headline {
   padding: 0px 15px;
-  margin: 0px;
+  margin:s 0px;
```
typo

ayamchiken committed:
Merge branch 'develop' of https://github.com/nathando/frappe into develop

Conflicts:
	frappe/public/build.json

Merge and update for signature

nathando committed:
Merge pull request #1 from ayamchiken/develop

minor fix for Signature

nathando committed:
Merge branch 'develop' into develop

nathando commented:
hi there, updated fixes for signature line thickness and print preview

nathando committed:
Make field uneditable after submitting

mhbu50 commented:
Is this feature will be add to V8?

nathando commented:
@netchampfaris will this be pushed as a feature soon ?

netchampfaris commented on frappe/public/js/frappe/form/control.js:
```diff
@@ -1706,7 +1706,119 @@ frappe.ui.form.ControlTable = frappe.ui.form.Control.extend({
+		}
+	},
+	load_pad: function() {
+        value = this.get_value();
```
use `var` declaration

netchampfaris commented on frappe/public/js/frappe/form/control.js:
```diff
@@ -1706,7 +1706,119 @@ frappe.ui.form.ControlTable = frappe.ui.form.Control.extend({
+	},
+	load_pad: function() {
+        value = this.get_value();
+        console.log("value", value, this.$pad);
```
remove console.log

netchampfaris commented on frappe/www/print.py:
```diff
@@ -148,8 +148,11 @@ def get_template_from_string():
 		"print_settings": frappe.get_doc("Print Settings")
 	}
-
+	
```
empty lines?

netchampfaris commented:
The Reset and Save buttons are very large, also we dont need a Save button. Only Reset will suffice.
Let's replace Reset button with a small Reset icon at the bottom right.

nathando commented:
@netchampfaris  Alright, let me push the changes this weekend. Let ship this feature!

nathando committed:
Remove console log and revert adding empty lines

nathando commented:
@netchampfaris: 
okay in the middle of the changes you requested but i have some problems:
- Use smaller button is ok but it's hard to press on mobile
- Use only reset button is ok as well but on desktop i can use blur() event to save, but on mobile I will need to do right before we save, because the blur event won't work so well on mobile.

Do you have any suggestions for these 2 problems ? Especially for number 2., if you have code reference on transforming the field right before the Save on a specific type of field, that'll be good.
Ex: How it looks on mobile:

![screen shot 2017-03-12 at 10 42 31 pm](https://cloud.githubusercontent.com/assets/1678810/23832712/5326f5b2-0775-11e7-80c1-cf9e1d5fc7e3.png)

netchampfaris commented:
@nathando Can you push your changes of buttons? I'll try looking for a solution for mobile.

nathando committed:
Update layout, remove edit mode/button, make save button compact. Refactor and use change event to trigger save.

nathando commented:
@netchampfaris i've got it working using the change event from jSignature. Haven't really tested on mobile but since it's based on when the stroke ends it should solve the problem. Quite a few code changes since we've removed the editing mode but i am happy with this version. Please help to double check.
There is a bug from the plugin jSignature though. When it reload the image to the canvas, it's thicker than the original... I've tested on the demo site and the result the same. For example:
Original
![screen shot 2017-03-18 at 10 38 09 pm](https://cloud.githubusercontent.com/assets/1678810/24073064/1888fa3e-0c2c-11e7-9dbf-7a451ae407d3.png)
Reimport (reloaded after pressing save)
![screen shot 2017-03-18 at 10 38 16 pm](https://cloud.githubusercontent.com/assets/1678810/24073065/1a530dbe-0c2c-11e7-967f-915043a9f48e.png)
This only causes problem if every time user add stroke to the existing canvas. In that case, previous strokes will get thicker. But most of the time ppl will reset first before trying. So i guess it's kinda pass.

rmehta commented:
Thanks @nathando 👏👏👏

Pushing after some minor fixes

albertusgeyser commented:
This Feature available yet? We need this to digitally sign documents...

paurosello commented:
This has been present since version 8 if I'm not wrong, check the signature fields

regards!

albertusgeyser commented:
Good Day

Thank You very much i never realised it . . . Great Job.

Just one other question? How can i put the signature in delivery note when  printed?
Is it a document design change i can do?

Thank You
Albertus Geyser

paurosello commented:
I have not used it yet... but I think it should be possible.

Have you tried adding a Custom Field?

On Oct 18, 2017 12:30 PM, "albertusgeyser" <notifications@github.com> wrote:

Good Day

Thank You very much i never realised it . . . Great Job.

Just one other question? How can i put the signature in delivery note when
printed?
Is it a document design change i can do?

Thank You
Albertus Geyser

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/frappe/frappe/pull/2433#issuecomment-337544028>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABFNgkJEcHBmYtnJSAld71Z8ODERhxUNks5stdNTgaJpZM4LIGxl>
.

nathando commented:
Hi, it has became a part of frappe. You can add a custom field with type Signature Field to your Delivery Note using Custom Field or Customize Form

Barbouze83 commented:
Hi,

Thank you Nathando and the other contributors for this functionality.

I can find the field to be added on a document (in the customization of the document) but can't find the controller to actually sign the document. Any advice where i can find it?

Thanks in advance,

Thomas

nathando commented:
@Barbouze83: Is there a reason why do you need to find it ? To know where is the javascript controller you can check file changes https://github.com/frappe/frappe/pull/2433/files
